### PR TITLE
[AMD] Fuse topk padded-token masking into a single Triton kernel

### PR DESCRIPTION
**Pull request by karverma-amd**
On AMD HIP the aiter MoE kernels cannot use topk_ids=-1 to drop CUDA-graph padding rows, so padded tokens are masked by zeroing their routing weights instead. The eager `arange + (>=) + boolean index_put_` path issues several launch-latency-bound kernels per call, twice per MoE layer.

## Modifications

1. Fused masking kernel: Added _fill_padded_rows, a single Triton kernel (one program per row; pad count read from device memory so it is CUDA/HIP-graph safe) that backs both _mask_topk_ids_padded_region and _zero_topk_weights_padded_region. Falls back to the eager path for non-contiguous tensors via _can_fuse_padded_region.

2. Removed a redundant pass: In _post_process_topk_ids, dropped the first HIP weight-zeroing call. The final pass already re-zeros after any shared-expert append/remap, so the earlier call was redundant (zeroing is idempotent) — halving the per-layer masking launches.

3. Opt-in full skip: Added the SGLANG_MORI_NO_PAD_MASK env flag (default off). When set, the HIP padded-token masking is skipped entirely, since padded rows are discarded downstream and the MoE combine is per-token. Gated off by default because it is a numerics-affecting change.

4. Tests: Added test/srt/layers/moe/test_topk_padded_region.py covering fused-vs-eager equivalence across shapes/dtypes, non-contiguous fallback, None pad-count no-op, CUDA-graph capture/replay, zeroing idempotence (justifying the redundant-call removal), and _post_process_topk_ids masking behavior with the flag on/off.

## Accuracy Tests

Unit tests: 8 tests / 37 subtests passing on MI355X.
GSM8k (DeepSeek-R1, MoRI EP, 1319 questions, 5-shot) with masking fully skipped (SGLANG_MORI_NO_PAD_MASK=1): 94.2%, Invalid 0.000 

Command Used -

1. Environment + launch the server:

export SGLANG_AITER_MOE=1
export AITER_FUSE_STAGE1_QUANT=1
export AITER_FUSE_GATEUP_NORM=0
export TORCH_NCCL_BLOCKING_WAIT=1 TRITON_ENABLE_PRELOAD_KERNELS=0
export SGLANG_USE_ROCM700A=1 SGLANG_ROCM_FUSED_DECODE_MLA=1 SGLANG_AITER_AR=1
export ROCM_QUICK_REDUCE_QUANTIZATION=INT4 ROCM_QUICK_REDUCE_MAX_SIZE_BYTES_MB=512
export SGLANG_AITER_MLA_PERSIST=1 SGLANG_USE_AITER=1 TORCH_BLAS_PREFER_HIPBLASLT=1
export PYTORCH_ALLOC_CONF=expandable_segments:True
export MORI_SHMEM_MODE=ISOLATION MORI_SHMEM_HEAP_SIZE=16G
export SGLANG_MORI_DISPATCH_DTYPE=auto SGLANG_MORI_COMBINE_DTYPE=fp8
export SGLANG_MORI_NO_PAD_MASK=1
export SGLANG_MORI_NUM_MAX_DISPATCH_TOKENS_PER_RANK=4096
python3 -m sglang.launch_server \
  --model-path /model/deepseek-r1/fp4_quantized/ \
  --tp-size 8 --ep-size 8 --dp-size 8 \
  --moe-a2a-backend mori --deepep-mode normal --moe-dense-tp-size 1 \
  --enable-dp-attention --enable-dp-lm-head \
  --mem-fraction-static 0.88 \
  --kv-cache-dtype fp8_e4m3 \
  --trust-remote-code \
  --triton-attention-num-kv-splits 2 \
  --disable-radix-cache --disable-chunked-prefix-cache \
  --host 0.0.0.0 --port 30000
2. Once the server is ready (/health_generate returns OK), run the eval:

python3 -m sglang.test.few_shot_gsm8k \
  --num-questions 1319 --num-shots 5 --parallel 1319 \
  --data-path /tmp/gsm8k_test.jsonl \
  --host http://127.0.0.1 --port 30000

## Speed Tests and Profiling

Before - 
image

After -

image

Results 

E2E + kernel-level perf comparison (DeepSeek-R1 MXFP4, MoRI-EP, TP8/EP8/DP8, MI355X)

E2E (8k/1k, ISL 8192 / OSL 1024): on par — within run-to-run variance. Interleaved 3-rep BEFORE/AFTER sweeps showed no consistent e2e throughput delta (session-to-session swings of ±15–20% on identical configs dominate). This is expected: server decode here is memory-bandwidth-bound and the masking kernels are captured inside the HIP graph, so the per-layer launch savings don't move aggregate tok/s above the noise floor. No regression.

Kernel-level (where this PR actually helps): the padded-region masking step — eager arange + (>=) + index_put_ vs the fused Triton kernel — with bit-identical outputs (equivalence verified):

image

1.3–1.4× per call (flat across rows → launch-latency-bound, exactly the motivation).

~2× / −50% per MoE layer (fusion + removal of the redundant second zeroing pass).

~1.05 ms saved per forward across the 58 MoE layers (rows=2048: ~2.08 → ~1.03 ms).

Fused kernel is HIP-graph-capturable; the eager index_put_ path is data-dependent and isn't.

This corroborates the before/after profiler traces above.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #27962681527](https://github.com/sgl-project/sglang/actions/runs/27962681527)
Latest PR Test (Extra): :x: [Run #27962681741](https://github.com/sgl-project/sglang/actions/runs/27962681741)

### CODE REVIEW

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]
Could you please move this test script to test/registered/moe and register the test for AMD's CI by adding 
`
register_amd_ci(est_time=xxx, stage="xxx", runner_config="xxx")
`
Otherwise, the test will not be run in our CI.

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]
Please also provide e2e performance comparison. Thanks.

**Comment by karverma-amd** [comment]
> Please also provide e2e performance comparison. Thanks.

<html>
<body>
<p data-pm-slice="0 0 []"><strong>E2E + kernel-level perf comparison</strong> (DeepSeek-R1 MXFP4, MoRI-EP, TP8/EP8/DP8, MI355X)</p><p><strong>E2E (8k/1k, ISL 8192 / OSL 1024): on par — within run-to-run variance.</strong>
Interleaved 3-rep BEFORE/AFTER sweeps showed no consistent e2e throughput delta (session-to-session swings of ±15–20% on identical configs dominate). This is expected: server decode here is memory-bandwidth-bound and the masking kernels are captured inside the HIP graph, so the per-layer launch savings don't move aggregate tok/s above the noise floor. <strong>No regression.</strong></p><p><strong>Kernel-level (where this PR actually helps):</strong> the padded-region masking step — eager <code>arange + (&gt;=) + index_put_</code> vs the fused Triton kernel — with <strong>bit-identical outputs</strong> (equivalence verified):</p>

image


<ul><li><p><strong>1.3–1.4× per call</strong> (flat across rows → launch-latency-bound, exactly the motivation).</p></li><li><p><strong>~2× / −50% per MoE layer</strong> (fusion + removal of the redundant second zeroing pass).</p></li><li><p><strong>~1.05 ms saved per forward</strong> across the 58 MoE layers (rows=2048: ~2.08 → ~1.03 ms).</p></li><li><p>Fused kernel is <strong>HIP-graph-capturable</strong>; the eager <code>index_put_</code> path is data-dependent and isn't.</p></li></ul><p>This corroborates the before/after profiler traces above.</p>
</body>
</html>

**Comment by karverma-amd** [comment]
> Could you please move this test script to test/registered/moe and register the test for AMD's CI by adding `register_amd_ci(est_time=xxx, stage="xxx", runner_config="xxx")` Otherwise, the test will not be run in our CI.

Done.

File: python/sglang/srt/layers/moe/topk.py [hunk 1]
```diff
@@ -1635,7 +1715,7 @@ def _post_process_topk_ids(
             topk_config,
         )
 
-    if _is_hip:
+    if _is_hip and not _skip_hip_pad_mask:
```
Comment (kkHuang-amd) [inline python/sglang/srt/layers/moe/topk.py:1718]: What's is the purpose of _skip_hip_pad_mask?

Look like the num_token_non_padded has checked the padding does or not.

What situation _skip_hip_pad_mask=true and num_token_non_padded is non-zero?

Comment (karverma-amd) [reply python/sglang/srt/layers/moe/topk.py:1718]: @kkHuang-amd ,   _skip_hip_pad_mask (env SGLANG_MORI_NO_PAD_MASK, default off) gates only the routing-weight zeroing — not the topk_ids masking. The padded topk_ids are still always set to a valid expert id (fill_value=0) above, unconditionally, so there's no -1/illegal-memory risk either way.

On your point about num_token_non_padded: yes, that's the padding boundary (rows >= num_token_non_padded are padding), and the mask helpers already no-op when it's None. _skip_hip_pad_mask is orthogonal to that — it doesn't decide whether there is padding, it decides whether we bother neutralizing the padded tokens' weights vs. letting the downstream combine drop them.

So _skip_hip_pad_mask=true with non-zero num_token_non_padded is the intended case: on the MoRI/DeepEP path there are CUDA-graph/dp-attention padded rows, but the combine is per-token and only sums the valid tokens — padded rows are discarded regardless of their weights, so zeroing them is redundant work (a launch-latency-bound kernel, twice per MoE layer). The flag skips that work.

It's default-off / numerics-affecting because correctness then relies on that downstream-discard guarantee.

File: python/sglang/srt/layers/moe/topk.py [hunk 2]
```diff
@@ -1172,6 +1246,8 @@ def _mask_topk_ids_padded_region(
         indices = torch.arange(0, topk_ids.shape[0], device=topk_ids.device)
         mask = (indices >= num_token_non_padded).unsqueeze(-1)
         topk_ids = torch.where(mask, torch.full_like(topk_ids, -1), topk_ids)
+    elif _can_fuse_padded_region(topk_ids):
+        _fill_padded_rows(topk_ids, num_token_non_padded, -1)
```
Comment (sogalin) [inline python/sglang/srt/layers/moe/topk.py]: Why do we force using -1 not to use "fill_value"?

Comment (karverma-amd) [reply python/sglang/srt/layers/moe/topk.py]: @sogalin -    Good catch — that hardcoded -1 is a leftover and should be fill_value. The fused branch is the path taken on HIP (and for non-int32 / non-CUDA cases), and the HIP caller passes fill_value=0 specifically because aiter MoE can't index topk_ids=-1 (negative index → illegal memory). Hardcoding -1 there is inconsistent with the else branch (which uses fill_value) and silently defeats that HIP intent.

Fixed to _fill_padded_rows(topk_ids, num_token_non_padded, fill_value). (The CUDA int32 + fill_value==-1 case still routes to the dedicated mask_topk_ids kernel above, so this only affects the fused/eager fallbacks, which now correctly honor fill_value.)

**Review by kkHuang-amd — APPROVED** [review]

**Comment by sogalin** [comment]
@amd-bot ci-status

**Comment by sogalin** [comment]
@karverma-amd , you registered this test case for CUDA, have you tried it on CUDA?

**Comment by karverma-amd** [comment]
> @karverma-amd , you registered this test case for CUDA, have you tried it on CUDA?

@sogalin -

Yes — ran test/registered/moe/test_topk_padded_region.py on CUDA (B200, lmsysorg/sglang:v0.5.13.post1-cu130 + PR #28084). 6 CUDA tests passed (37 subtests), including CUDA graph capture/replay. The 2 TestPostProcessPaddedMaskingHip tests are HIP-only by design and skip on CUDA.

**Review by hubertlu-tw — APPROVED** [review]
LGTM
